File: gfg/april2024/c++/04_sum_of_all_substrings_of_number.cpp
// CODE

#define MOD 1000000007
class Solution {
    public:
    long long sumSubstrings(string s) {
        int n = s.size();
        long long res;
        vector<long long> dp(n, 0);
        
        dp[0] = s[0] - '0';
        res = dp[0];
        
        for(int i = 1; i < n; i++) {
            dp[i] = ((dp[i - 1] * 10) % MOD + ((s[i] - '0') * (i + 1)) % MOD) % MOD;
            res = (res + dp[i]) % MOD;
        }
        
        return res;
    }
};

File: gfg/2024_4_april/python/04_sum_of_all_substrings_of_number.py
# CODE

MOD = 10**9 + 7
class Solution:
    def sumSubstrings(self, s: str) -> int:
        n = len(s)
        dp = [0] * n
        dp[0] = int(s[0])
        res = dp[0]
        for i in range(1, n):
            dp[i] = ((dp[i - 1] * 10) % MOD + ((int(s[i]) * (i + 1)) % MOD)) % MOD
            res = (res + dp[i]) % MOD
        return res